src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h

```diff
@@ -0,0 +1,1637 @@
+// -*- mode:C++; tab-width:8; c-basic-index:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <iostream>
+#include <string>
+
+#include "include/byteorder.h"
+#include "include/denc.h"
+
+#include "crimson/common/layout.h"
+#include "crimson/common/fixed_kv_node_layout.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_types.h"
+
+#define BLOCK_SIZE 4096
+namespace crimson::os::seastore::omap_manager {
+
+template <
+  typename Meta,
+  typename MetaInt,
+  bool VALIDATE_INVARIANTS=true> class StringKVInnerNodeLayout;
+
+template <
+  typename Meta,
+  typename MetaInt,
+  bool VALIDATE_INVARIANTS=true> class StringKVLeafNodeLayout;
+
+
+/**
+  * copy_from_foreign
+  *
+  * Copy from another node entries to this node.
+  * [from_src, to_src) is another node entry range.
+  * tgt is this node entry to copy to.
+  * tgt and from_src must be from different nodes.
+  * from_src and to_src must be in the same node.
+  */
+template <typename iterator, typename const_iterator>
+static void copy_from_foreign(
+  iterator tgt,
+  const_iterator from_src,
+  const_iterator to_src) {
+  assert(tgt->node != from_src->node);
+  assert(to_src->node == from_src->node);
+  if (from_src == to_src)
+    return;
+
+  auto to_copy = from_src->get_right_ptr_end() - to_src->get_right_ptr_end();
```

Comment:
Here's an example re my previous comment on not using unsigned ints for offsets...